pkg/kubectl/cmd/run.go

```diff
@@ -306,45 +345,100 @@ func contains(resourcesList map[string]*unversioned.APIResourceList, resource un
 	return false
 }
 
-func waitForPodRunning(c *client.Client, pod *api.Pod, out io.Writer, quiet bool) (status api.PodPhase, err error) {
+// waitForPod watches the given pod until the exitCondition is true. Each two seconds
+// the tick function is called e.g. for progress output.
+func waitForPod(c *client.Client, ns, name string, exitCondition func(*api.Pod) bool, tick func(*api.Pod)) (*api.Pod, error) {
+	pod, err := c.Pods(ns).Get(name)
+	if err != nil {
+		return nil, err
+	}
+	if exitCondition(pod) {
+		return pod, nil
+	}
+
+	tick(pod)
+
+	w, err := c.RESTClient.Get().
+		Prefix("watch").
+		Resource("pods").
+		Namespace(ns).
+		Name(name).
+		VersionedParams(&api.ListOptions{
+			ResourceVersion: pod.ResourceVersion,
+		}, api.ParameterCodec).
+		Watch()
+	if err != nil {
+		return nil, err
+	}
+	defer w.Stop()
+
+	t := time.NewTicker(2 * time.Second)
+	defer t.Stop()
+
 	for {
-		pod, err := c.Pods(pod.Namespace).Get(pod.Name)
-		if err != nil {
-			return api.PodUnknown, err
+		select {
```

Comment:
This should be using `watch.Wait`, since you're effectively doing the same thing.  See all the client conditions for this, and add the proper ones there or reuse the existing.